File: 242-valid-anagram/valid-anagram.py
class Solution(object):
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        chars_and_appearances = {}
        
        for i in s:
            if i in chars_and_appearances:
                chars_and_appearances[i] += 1
            else:
                chars_and_appearances[i] = 1
        
        for i in t:
            if i in chars_and_appearances:
                chars_and_appearances[i] -= 1
                if chars_and_appearances[i] < 0:
                    return False
            else:
                return False
        
        for i in chars_and_appearances.values():
            if i != 0:
                return False
        return True

File: 242-valid-anagram/valid-anagram.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        map<int, int> counter;
        if(s.size() != t.size())
            return false;
        for(char i : s)
        {
            if(counter.find(i) != counter.end())
                counter.find(i)->second++;
            else
                counter[i] = 1;
        }
        for(char i : t)
        {
            auto curr = counter.find(i);
            if(curr == counter.end() || curr->second == 0)
                return false;
            curr->second--;
        }
        for(auto i : counter)
        {
            if(i.second != 0)
                return false;
        }
        return true;
    }
};